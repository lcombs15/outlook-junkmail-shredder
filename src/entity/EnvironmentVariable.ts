export enum EnvironmentVariableName {
    CLIENT_ID_FILE = 'CLIENT_ID_FILE',
    DISCORD_URL_FILE = 'DISCORD_URL_FILE',
    AUTH_DIRECTORY = 'AUTH_DIRECTORY',
    SUMMARY_FILE = 'SUMMARY_FILE'
}
