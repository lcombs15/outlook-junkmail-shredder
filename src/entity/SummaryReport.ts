import Email from "./email";

export interface EmailSummary {
    [email: string]: {
        timestamps: Array<string>;
        total: number;
    }
}

export interface GroupEmailSummary {
    details: {
        [reason: string]: {
            summary: EmailSummary;
            total: number;
        }
    }
    total: number;
}

export interface SummaryReport {
    deleted: GroupEmailSummary,
    ignored: GroupEmailSummary;
}