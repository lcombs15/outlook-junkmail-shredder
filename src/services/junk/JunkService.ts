import Email from "../../entity/email";
import {JunkStrategy} from "./strategy/JunkStrategy";
import {BogusNewsletterStrategy} from "./strategy/BogusNewsletterStrategy";
import {BogusOnMicrosoftStrategy} from "./strategy/BogusOnMicrosoftStrategy";
import {LooksLikeMyEmailStrategy} from "./strategy/LooksLikeMyEmailStrategy";
import {IllegalTopLevelDomainStrategy} from "./strategy/IllegalTopLevelDomainStrategy";
import {IllegalDomainStrategy} from "./strategy/IllegalDomainStrategy";
import {DontKnowMyNameDotInfo} from "./strategy/DontKnowMyNameDotInfo";

export interface JunkEvaluation {
    isJunk: boolean,
    reason: string,
}

export class JunkService {

    constructor(private strategies: Array<JunkStrategy> = [
        new BogusNewsletterStrategy(),
        new IllegalTopLevelDomainStrategy(),
        new IllegalDomainStrategy(),
        new DontKnowMyNameDotInfo(),
        new BogusOnMicrosoftStrategy(),
        new LooksLikeMyEmailStrategy()
    ]) {
    }

    public evaluate(email: Email): JunkEvaluation {
        const match = this.strategies.find(strat => strat.appliesTo(email));

        return {
            isJunk: !!match,
            reason: match?.getReason(email) || 'No matching junk filter'
        }
    }
}
