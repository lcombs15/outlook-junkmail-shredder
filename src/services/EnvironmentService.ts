import {EnvironmentVariableName} from "../entity/EnvironmentVariable";
import * as fs from "node:fs";

export class EnvironmentService {

    public constructor(private sysEnv = process.env) {
    }

    public getValue(key: EnvironmentVariableName): string | null {
        const value = this.sysEnv[key.toString()];

        if (!value) {
            console.error('Missing environment variable: ', key.toString());
        }

        return value || null;
    }

    public getRequiredValue(key: EnvironmentVariableName): string {
        const val = this.getValue(key);

        if (val == null) {
            throw new Error(`Missing environment variable: ${key.toString()}`)
        }

        return val;
    }

    public getValueFromFile(key: EnvironmentVariableName): string | null {
        const fileName = this.getValue(key);

        if (!fileName) {
            return null;
        }

        return fs.readFileSync(fileName).toString()
    }
}
