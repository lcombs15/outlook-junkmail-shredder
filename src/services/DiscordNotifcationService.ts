import {EnvironmentService} from "./EnvironmentService";
import {EnvironmentVariableName} from "../entity/EnvironmentVariable";
import Email, {EmailAddress} from "../entity/email";
import {JunkEvaluation} from "./junk/JunkService";

export class DiscordNotificationService {
    private readonly url: string;

    constructor(environmentService: EnvironmentService, private httpClient = fetch) {
        this.url = environmentService.getValueFromFile(EnvironmentVariableName.DISCORD_URL_FILE) || 'no discord file';
    }

    private emailToString(address: EmailAddress): string {
        if (!address?.emailAddress) {
            return '';
        }

        return `${address.emailAddress.name} [${address.emailAddress.address}]`
    }

    public async sendEmailMessage(messageTitle: string, emails: Array<[Email, JunkEvaluation]>) {
        return this.sendMessage(messageTitle, emails.map(([email, junkEvaluation]) => {
            return {
                subject: email.subject,
                from: this.emailToString(email.from),
                sender: this.emailToString(email.sender),
                to: (email.toRecipients || []).map(this.emailToString).join(','),
                description: junkEvaluation.reason
            }
        }));
    }

    public async sendMessage(messageTitle: string, embeds: Array<Record<string, string>>): Promise<void> {
        if (!this.url) {
            console.warn('Discord URL not provided, skipping notification');
            return Promise.resolve();
        }

        const payload = {
            content: messageTitle,
            embeds: embeds.map(embed => {
                return {
                    fields: Object.entries(embed)
                        .filter(([_, value]) => value)
                        .map(([key, value]) => {
                            return {
                                name: key,
                                value: value.substring(0, 1024),
                            }
                        })
                }
            })
        };

        try {
            const response = await this.httpClient(this.url, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify(payload)
            });
            if (!response.ok) {
                console.error(JSON.stringify(payload));
                throw new Error(`Discord notification failed: ${response.status} ${response.statusText}`);
            }
            console.log('Discord notification sent successfully', embeds);
        } catch (error) {
            console.error('Error sending Discord notification:', error);
            throw error;
        }
    }
}
