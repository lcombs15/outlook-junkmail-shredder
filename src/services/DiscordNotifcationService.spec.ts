import {DiscordNotificationService} from "./DiscordNotifcationService";
import {EnvironmentService} from "./EnvironmentService";
import {mock, MockProxy} from 'jest-mock-extended';
import Email from "../entity/email";
import {JunkEvaluation} from "./junk/JunkService";
import {EnvironmentVariableName} from "../entity/EnvironmentVariable";

const mockEvaluation: [Email, JunkEvaluation] = [
    {
        id: 'aaa-bbb-ccc',
        receivedDateTime: '2025 Mar 7',
        toRecipients: [{
            emailAddress: {
                name: 'Steve Jobs',
                address: 'steve@apple.com'
            }
        }],
        bccRecipients: [],
        ccRecipients: [],
        subject: '',
        from: {
            emailAddress: {
                name: 'Post Master From',
                address: 'post+from@postmaster.com'
            }
        },
        body: {
            content: '<html></html>',
            contentType: 'html'
        },
        sender: {
            emailAddress: {
                name: 'Post Master Sender',
                address: 'post+sender@postmaster.com'
            }
        }
    },
    {
        isJunk: false,
        reason: 'test data is not junk!'
    }
]

describe('DiscordNotificationService', () => {
    let service: DiscordNotificationService;

    let http = jest.fn();

    let environmentService: MockProxy<EnvironmentService> = mock<EnvironmentService>();

    const expectedUrl = "mydiscord.com/the-server/stuff"

    beforeEach(() => {
        http.mockClear();

        environmentService.getValueFromFile
            .calledWith(EnvironmentVariableName.DISCORD_URL_FILE)
            .mockReturnValue(expectedUrl);

        service = new DiscordNotificationService(environmentService, http);
    })

    it('should send a message to discord', async () => {
        http.mockReturnValue({
            ok: true
        })

        await service.sendEmailMessage('msg 2 discord', [mockEvaluation]);

        const [url, body] = http.mock.calls[0];

        expect(url).toBe(expectedUrl);
        expect(body).toMatchSnapshot()
    });

    it('should truncate long embeds', async () => {
        http.mockReturnValue({
            ok: true
        })

        const myIncludedData = new Array(2000).fill('abc').join();

        await service.sendMessage('msg 2 discord trunc', [{myIncludedData}]);

        const [url, request] = http.mock.calls[0];

        expect(url).toBe(expectedUrl);
        const {embeds} = JSON.parse(request.body)
        expect(embeds).toHaveLength(1);
        expect(embeds[0].fields).toHaveLength(1);

        const mySingleEmbed = embeds[0].fields[0];
        expect(mySingleEmbed.name).toBe('myIncludedData');
        expect(mySingleEmbed.value).toHaveLength(1024);
        expect(myIncludedData.length).toBeGreaterThan(1024);
    });
})
