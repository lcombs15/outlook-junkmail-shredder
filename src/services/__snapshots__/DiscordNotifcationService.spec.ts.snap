// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`DiscordNotificationService should send a message to discord 1`] = `
{
  "body": "{"content":"msg 2 discord","embeds":[{"fields":[{"name":"from","value":"Post Master From [post+from@postmaster.com]"},{"name":"sender","value":"Post Master Sender [post+sender@postmaster.com]"},{"name":"to","value":"Steve Jobs [steve@apple.com]"},{"name":"description","value":"test data is not junk!"}]}]}",
  "headers": {
    "Content-Type": "application/json",
  },
  "method": "POST",
}
`;
